File: 0079-word-search/0079-word-search.java
class Solution {
    public int[][] side;
    public int c, r;
    
    public boolean exist(char[][] board, String word) {
        side = new int[board.length][board[0].length];
        r = board.length;
        c = board[0].length;
        for (int i = 0;i<r;i++)
        {
            for (int j = 0;j<c;j++)
            {
                if (doesExist(board, word, 0, i, j) == true)
                {
                    return true;
                }
            }
        }
        return false;
    }
    
    public boolean doesExist(char[][] board, String word, int wordCount, int rc, int cc)
    {
        if (isSafe(rc, cc) == false || wordCount == word.length() || board[rc][cc] != word.charAt(wordCount))
        {
            return false;
        }
        if (board[rc][cc] == word.charAt(wordCount) && wordCount == word.length() - 1)
        {
            return true;
        }
        
        side[rc][cc] = 1;
        boolean exists = doesExist(board, word, wordCount+1, rc+1, cc) || doesExist(board, word, wordCount+1, rc, cc+1) || doesExist(board, word, wordCount+1, rc-1, cc) || doesExist(board, word, wordCount+1, rc, cc-1);
        
        side[rc][cc] = 0;
                
        return exists;
    }
    
    public boolean isSafe(int rc, int cc)
    {
        return (rc > -1 & rc < r && cc > -1 && cc < c && side[rc][cc] == 0);
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])
        found = False

        def dp(i, j, ind):
            nonlocal temp, found
            
            if ind >= len(word):
                found = True
                return 
            if found == True or i >= rows or i < 0 or j >= cols or j < 0 or temp[i][j] == True or board[i][j] != word[ind]:
                return
            temp[i][j] = True
            ind += 1
            dp(i, j + 1, ind)
            dp(i + 1, j, ind)
            dp(i, j - 1, ind)
            dp(i - 1, j, ind)
            temp[i][j] = False
            return

        for i in range(rows):
            for j in range(cols):
                temp = [[False for _ in range(cols)] for _ in range(rows)]
                if found == False and board[i][j] == word[0]:
                    dp(i, j, 0)

        return found